File: BOJ/Questions/Q14939.cpp
//Question No: 14939
//Title: 불 끄기
//Tier: Platinum IV
#include <bits/stdc++.h>
using namespace std;
#define MAX 10
#define INF 1e9

bool arr[15][15];
int dx[] = { 0,0,1,-1 };
int dy[] = { 1,-1,0,0 };
int ans = INF;

bool outrange(int x, int y) {
    return (x < 0 || x >= MAX || y < 0 || y >= MAX);
}

void toggle(int x, int y, bool tmp_arr[][15]) {
    for (int i = 0; i < 4; i++) {
        int nx = x + dx[i];
        int ny = y + dy[i];
        if (!outrange(nx, ny)) tmp_arr[nx][ny] = !tmp_arr[nx][ny];
    }
    tmp_arr[x][y] = !tmp_arr[x][y];
}

bool islight(bool tmp_arr[][15]) {
    for (int i = 0; i < MAX; i++)
        for (int j = 0; j < MAX; j++)
            if (tmp_arr[i][j]) return true;
    return false;
}

void init(bool t_arr1[][15], bool t_arr2[][15], bool t_arr[][15]) {
    for (int i = 0; i < MAX; i++)
        for (int j = 0; j < MAX; j++) {
            t_arr2[i][j] = t_arr[i][j];
            t_arr1[i][j] = t_arr[i][j];
        }
}

void firstLine(int x, int sum, bool t_arr[][15]) {
    if (x == MAX) {
        bool t_arr3[15][15] = { 0 };
        for (int i = 0; i < MAX; i++)
            for (int j = 0; j < MAX; j++)
                t_arr3[i][j] = t_arr[i][j];

        for (int i = 1; i < MAX; i++) {
            for (int j = 0; j < MAX; j++) {
                if (t_arr3[i - 1][j]) {
                    toggle(i, j, t_arr3);
                    sum++;
                }
            }
        }

        if (!islight(t_arr3))  ans = min(ans, sum);
        return;
    }

    bool t_arr1[15][15] = { 0 };
    bool t_arr2[15][15] = { 0 };
    init(t_arr1, t_arr2, t_arr);

    firstLine(x + 1, sum, t_arr1);

    toggle(0, x, t_arr2);
    firstLine(x + 1, sum + 1, t_arr2);
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(NULL); cout.tie(NULL);

    for (int i = 0; i < MAX; i++) {
        for (int j = 0; j < MAX; j++) {
            char c; cin >> c;
            if (c == 'O') arr[i][j] = 1;
        }
    }

    firstLine(0, 0, arr);

    if (ans == INF) cout << -1;
    else cout << ans;
}

File: BOJ/Questions/Q14939.java
//Question No: 14939
//Title: 불 끄기
//Tier: Platinum IV
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class Main {
    private static final int MAX = 10;
    private static final int INF = 1000000000;
    private static boolean[][] arr = new boolean[MAX][MAX];
    private static int[] dx = {0, 0, 1, -1};
    private static int[] dy = {1, -1, 0, 0};
    private static int minSteps = INF;

    public static void main(String[] args) throws IOException {
        BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
        for (int i = 0; i < MAX; i++) {
            String line = reader.readLine();
            for (int j = 0; j < MAX; j++) {
                arr[i][j] = line.charAt(j) == 'O';
            }
        }
        solveFirstLine(0, 0, arr);
        System.out.println(minSteps == INF ? -1 : minSteps);
    }

    private static boolean isOutOfRange(int x, int y) {
        return x < 0 || x >= MAX || y < 0 || y >= MAX;
    }

    private static void toggle(int x, int y, boolean[][] tempArr) {
        for (int i = 0; i < 4; i++) {
            int nx = x + dx[i];
            int ny = y + dy[i];
            if (!isOutOfRange(nx, ny)) {
                tempArr[nx][ny] = !tempArr[nx][ny];
            }
        }
        tempArr[x][y] = !tempArr[x][y];
    }

    private static boolean isLightOn(boolean[][] tempArr) {
        for (int i = 0; i < MAX; i++) {
            for (int j = 0; j < MAX; j++) {
                if (tempArr[i][j]) {
                    return true;
                }
            }
        }
        return false;
    }

    private static void copyArray(boolean[][] source, boolean[][] dest1, boolean[][] dest2) {
        for (int i = 0; i < MAX; i++) {
            for (int j = 0; j < MAX; j++) {
                dest1[i][j] = source[i][j];
                dest2[i][j] = source[i][j];
            }
        }
    }

    private static void solveFirstLine(int x, int sum, boolean[][] tempArr) {
        if (x == MAX) {
            boolean[][] tempArrCopy = new boolean[MAX][MAX];
            for (int i = 0; i < MAX; i++) {
                for (int j = 0; j < MAX; j++) {
                    tempArrCopy[i][j] = tempArr[i][j];
                }
            }

            for (int i = 1; i < MAX; i++) {
                for (int j = 0; j < MAX; j++) {
                    if (tempArrCopy[i - 1][j]) {
                        toggle(i, j, tempArrCopy);
                        sum++;
                    }
                }
            }

            if (!isLightOn(tempArrCopy)) {
                minSteps = Math.min(minSteps, sum);
            }
            return;
        }

        boolean[][] tempArr1 = new boolean[MAX][MAX];
        boolean[][] tempArr2 = new boolean[MAX][MAX];
        copyArray(tempArr, tempArr1, tempArr2);

        solveFirstLine(x + 1, sum, tempArr1);

        toggle(0, x, tempArr2);
        solveFirstLine(x + 1, sum + 1, tempArr2);
    }
}
